### PR TITLE
Harmony 1764 2

### PR DESCRIPTION
**Pull request by indiejames**
## Jira Issue ID
HARMONY-1764

## Description
We use SQS queues to communicate between the main harmony executable and internal services. SQS queue messages are limited to 256K. When a user specifies many variables the data operation can exceed this size. This change uses an in-memory cache for the data operation so it does not need to be passed on the queues. 

## Local Test Steps
Testing this change involves a lot of steps.
1. Check out this branch
2. Run the normal dev services and test with a standard query
3. Rebuild the service images
```
npm run build-all
```
4. Change your .env to run harmony in a box
5. Run harmony in a box
```
./bin/bootstrap-harmony
```
6. Run some queries to test things out

To test out a full request with many variables you will need to rebuild a service (like trajectory-subsetter) with my harmony-1764-2 branch of harmony-service-lib-py. I can provide a query for that if needed.

## PR Acceptance Checklist
* [x] Acceptance criteria met
* [x] Tests added/updated (if needed) and passing
* [ ] Documentation updated (if needed)

### CODE REVIEW

**Review by ygliuvt — APPROVED** [review]

File: services/harmony/app/backends/workflow-orchestration/work-item-polling.ts [hunk 1]
```diff
@@ -12,6 +16,36 @@ export type WorkItemData = {
   maxCmrGranules?: number
 };
 
+
+/**
+ * Function to fetch the operation for a work-item when it is not in the cache
+ *
+ * @param jobID - the id of the job
```
[the remark this answers is not in the document]
Comment (indiejames) [reply services/harmony/app/backends/workflow-orchestration/work-item-polling.ts]: Done

File: services/harmony/env-defaults [hunk 2]
```diff
@@ -123,6 +123,9 @@ MAX_POST_FILE_SIZE=2000000000
 # Maximum number of multipart parts to accept when providing a shapefile
 MAX_POST_FILE_PARTS=100
 
+# Maximum size (in bytes) of the cache for data operations
+MAX_DATA_OPERATION_CACHE_SIZE=2000000000
```
[the remark this answers is not in the document]
Comment (indiejames) [reply services/harmony/env-defaults:127]: Done.

File: services/service-runner/app/service/service-runner.ts [hunk 3]
```diff
@@ -257,6 +258,20 @@ export async function runServiceFromPull(workItem: WorkItemRecord, workItemLogge
...
 
+      // use a temporary file on the shared volume mount to pass the operation if the operation is large
+      const operationJsonFile = '/tmp/operation.json';
+      const operationJson = JSON.stringify(operation);
+      let operationCommandLine = '--harmony-input';
+      let operationCommandLineValue = operationJson;
+      // 262144 is the max SQS message size, so any operation + other stuff we add that is
+      //  bigger than that is considered BIG and therefore requires special handling. In this case
+      // we use a file to pass the operation to harmony-service-lib instead of a command line argument.
+      if (operationJson.length > 100000) {
```
[the remark this answers is not in the document]
Comment (indiejames) [reply services/service-runner/app/service/service-runner.ts:269]: Done.

File: services/service-runner/app/service/service-runner.ts [hunk 4]
```diff
@@ -257,6 +258,20 @@ export async function runServiceFromPull(workItem: WorkItemRecord, workItemLogge
         resolve({ error: `Worker timed out after ${workerTimeout / 1000.0} seconds` });
       }, workerTimeout);
 
+      // use a temporary file on the shared volume mount to pass the operation if the operation is large
```
[the remark this answers is not in the document]
Comment (indiejames) [reply services/service-runner/app/service/service-runner.ts:261]: Done.

**Review by chris-durbin — APPROVED** [review]

**Pull request merged**